src/run.rs

```
@@ -0,0 +1,202 @@
+            let new_chan = send_chan.clone();
+            std_err_handle = Some(thread::spawn(move || {
+                read_stream(&new_name, new_chan, &mut buffered_stdout, false);
+            }));
+        }
+
+        if let Some(handle) = std_out_handle {
+            let _ = handle.join();
+        }
+
+        if let Some(handle) = std_err_handle {
+            let _ = handle.join();
+        }
+
+        let exit_status = shared_handle.lock().unwrap().wait();
+        match exit_status {
+            Ok(status) => {
+                let _ = send_chan.send(OutputMessage {
+                    name: command_name.clone(),
+                    message: OutputMessagePayload::Done(status.code()),
+                });
+
+                match options.restart {
+                    RestartOptions::Continue => {
+                        return (command_name, Some(status));
+                    }
+                    RestartOptions::Restart => {
+                        if status.success() {
+                            return (command_name, Some(status));
+                        }
+                    }
+                    RestartOptions::Kill => {
+                        if !status.success() {
+                            let _ = kill_trigger.initiate_kill();
+                        }
+                        return (command_name, Some(status));
+                    }
+                };
+            }
+            Err(e) => {
+                let _ = send_chan.send(OutputMessage {
+                    name: command_name.clone(),
+                    message: OutputMessagePayload::Error(e),
+                });
+                return (command_name, None);
+            }
+        }
+    })
+}
+
+fn kill_thread(kill_trigger: &kill_barrier::KillBarrier, child: Arc<Mutex<process::Child>>) {
+    let _ = kill_trigger.wait();
+
+    let lock_res = child.lock();
+    if let Ok(mut locked_child) = lock_res {
+        let _ = locked_child.kill();
+    }
```

What happens if it cannot acquire a lock on `child`? Does it not kill it then?

---

I failed lock acquisition comes from a poisoned lock. In rust land, that means that some other thread has panicked while holding that lock. 
In that case, I let that panic take charge so as to not pollute the stack trace

---

Ah, I forgot about the poisoned lock issue. Thanks for the explanation!